File: salary-calculator-client/src/app/salary.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface SalaryCalculationRequest {
  positionPercentage: number;
  professionalLevel: string;
  managementLevel: string;
  yearsOfExperience: number;
  isEntitledToBonus: boolean;
  group: string | null;
}

export interface SalaryCalculationResponse {
  baseSalaryPerHour: number;
  baseSalary: number;
  seniorityBonusAmount: number;
  seniorityBonusPercentage: number;
  additionalWorkBonus: number;
  baseSalaryBeforeRaise: number;
  raisePercentage: number;
  raiseAmount: number;
  finalSalary: number;
}

@Injectable({ providedIn: 'root' })
export class SalaryService {
  private baseUrl = 'https://localhost:7116/api/Salary';

  constructor(private http: HttpClient) {}

  calculateSalary(
    data: SalaryCalculationRequest | any
  ): Observable<SalaryCalculationResponse> {
    const params = new HttpParams({
      fromObject: {
        JobPercentage: data.positionPercentage,
        JobRole: data.professionalLevel,
        ManagementLevel: data.managementLevel,
        ExperienceYears: data.yearsOfExperience,
        IsEligibleForAdditionalWork: data.isEntitledToBonus,
        AdditionalWorkGroup: data.group,
      },
    });

    return this.http.get<SalaryCalculationResponse>(this.baseUrl, { params });
  }
}

File: calculator-backend/Services/SalaryService.cs
using calculator_backend.Models;
using calculator_backend.Services;
using MySalaryCalculatorApi.Enums;

namespace MySalaryCalculatorApi.Services
{
    public class SalaryService : ISalaryService
    {
        private const double HoursPerMonthForFullTime = 170;
        private const double BaseSalaryPerHourForJunior = 100;
        private const double BaseSalaryPerHourForSenior = 120;
        private const double SalaryIncreasePerManagementLevel = 20;
        private const double SeniorityBonusPercentage = 1.25;
        private const double AdditionalWorkBonusGroupA = 1;
        private const double AdditionalWorkBonusGroupB = 0.5;

        public SalaryCalculationResult CalculateSalary(SalaryCalculationModel model)
        {
            if (model.ExperienceYears < 0 || model.JobPercentage <= 0 || model.JobPercentage > 100)
            {
                throw new ArgumentException("Invalid input values. Please ensure all inputs are logical and valid.");
            }

            // Calculate base salary
            double baseSalaryPerHour = model.JobRole == JobRoleEnum.Junior ? BaseSalaryPerHourForJunior : BaseSalaryPerHourForSenior;
            baseSalaryPerHour += (int)model.ManagementLevel * SalaryIncreasePerManagementLevel;
            double baseSalary = baseSalaryPerHour * HoursPerMonthForFullTime * (model.JobPercentage / 100);

            // Calculate seniority bonus
            double seniorityBonus = baseSalary * model.ExperienceYears * SeniorityBonusPercentage / 100;

            // Calculate additional work bonus
            double additionalWorkBonus = 0;
            if (model.IsEligibleForAdditionalWork)
            {
                additionalWorkBonus = baseSalary * (model.AdditionalWorkGroup == AdditionalWorkGroupEnum.GroupA ? AdditionalWorkBonusGroupA : AdditionalWorkBonusGroupB) / 100;
            }

            // Calculate base salary before raise
            double baseSalaryBeforeRaise = baseSalary + seniorityBonus + additionalWorkBonus;

            // Calculate raise percentage based on base salary and management level
            double raisePercentage =
              baseSalaryBeforeRaise <= 20000 ? 1.5 :
              baseSalaryBeforeRaise <= 30000 ? 1.25 :
              1;
            raisePercentage += (int)model.ManagementLevel * 0.1;

            // Calculate raise amount
            double raiseAmount = baseSalaryBeforeRaise * raisePercentage / 100;

            // Calculate final salary
            double finalSalary = baseSalaryBeforeRaise + raiseAmount;

            // Create and return SalaryCalculationResult object
            return new SalaryCalculationResult
            {
                BaseSalary = baseSalary,
                SeniorityBonusPercentage = SeniorityBonusPercentage,
                SeniorityBonusAmount = seniorityBonus,
                AdditionalWorkBonus = additionalWorkBonus,
                BaseSalaryBeforeRaise = baseSalaryBeforeRaise,
                RaisePercentage = raisePercentage,
                RaiseAmount = raiseAmount,
                FinalSalary = finalSalary,
            };
        }
    }
}
